File: microservices/categories-service/src/main/java/com/categoriesservice/categories/service/CategoryService.java
package com.categoriesservice.categories.service;

import com.categoriesservice.categories.dto.CategoryDTO;
import com.categoriesservice.categories.model.Category;
import com.categoriesservice.categories.repository.CategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Category Service - Lógica de negocio para gestión de categorías
 */
@Service
public class CategoryService {
    
    @Autowired
    private CategoryRepository categoryRepository;
    
    public List<CategoryDTO> listarTodas() {
        return categoryRepository.findAll().stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }
    
    public CategoryDTO guardar(CategoryDTO categoryDTO) {
        Category category = convertToEntity(categoryDTO);
        Category savedCategory = categoryRepository.save(category);
        return convertToDTO(savedCategory);
    }
    
    public Optional<CategoryDTO> obtenerPorId(Long id) {
        return categoryRepository.findById(id)
                .map(this::convertToDTO);
    }
    
    public List<CategoryDTO> obtenerPorUsuario(Long userId) {
        return categoryRepository.findByUserId(userId).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }
    
    public List<CategoryDTO> buscarPorNombre(String nombre) {
        return categoryRepository.findByNameContainingIgnoreCase(nombre).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }
    
    public Optional<CategoryDTO> actualizar(Long id, CategoryDTO categoryDTO) {
        return categoryRepository.findById(id)
                .map(category -> {
                    category.setName(categoryDTO.getName());
                    category.setColor(categoryDTO.getColor());
                    category.setUserId(categoryDTO.getUserId());
                    Category updated = categoryRepository.save(category);
                    return convertToDTO(updated);
                });
    }
    
    public boolean eliminar(Long id) {
        if (categoryRepository.existsById(id)) {
            categoryRepository.deleteById(id);
            return true;
        }
        return false;
    }
    
    private CategoryDTO convertToDTO(Category category) {
        CategoryDTO dto = new CategoryDTO();
        dto.setId(category.getId());
        dto.setName(category.getName());
        dto.setColor(category.getColor());
        dto.setUserId(category.getUserId());
        return dto;
    }
    
    private Category convertToEntity(CategoryDTO dto) {
        Category category = new Category();
        category.setId(dto.getId());
        category.setName(dto.getName());
        category.setColor(dto.getColor());
        category.setUserId(dto.getUserId());
        return category;
    }
}

File: microservices/categories-service/src/main/kotlin/com/categoriesservice/categories/service/CategoryService.kt
package com.categoriesservice.categories.service

import com.categoriesservice.categories.dto.CategoryDTO
import com.categoriesservice.categories.dto.toDTO
import com.categoriesservice.categories.dto.toEntity
import com.categoriesservice.categories.repository.CategoryRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class CategoryService(
    private val repository: CategoryRepository
) {
    
    fun findAll(): List<CategoryDTO> = 
        repository.findAll().map { it.toDTO() }
    
    fun findAllActive(): List<CategoryDTO> = 
        repository.findByActiveTrue().map { it.toDTO() }
    
    fun findById(id: Long): CategoryDTO = 
        repository.findById(id)
            .map { it.toDTO() }
            .orElseThrow { NoSuchElementException("Category not found with id: $id") }
    
    fun create(dto: CategoryDTO): CategoryDTO {
        if (repository.existsByName(dto.name)) {
            throw IllegalArgumentException("Category with name '${dto.name}' already exists")
        }
        val entity = dto.toEntity()
        val saved = repository.save(entity)
        return saved.toDTO()
    }
    
    fun update(id: Long, dto: CategoryDTO): CategoryDTO {
        val existing = repository.findById(id)
            .orElseThrow { NoSuchElementException("Category not found with id: $id") }
        
        // Verificar nombre duplicado si se está cambiando
        if (existing.name != dto.name && repository.existsByName(dto.name)) {
            throw IllegalArgumentException("Category with name '${dto.name}' already exists")
        }
        
        existing.name = dto.name
        existing.description = dto.description
        existing.active = dto.active
        
        val updated = repository.save(existing)
        return updated.toDTO()
    }
    
    fun delete(id: Long) {
        if (!repository.existsById(id)) {
            throw NoSuchElementException("Category not found with id: $id")
        }
        repository.deleteById(id)
    }
    
    fun toggleActive(id: Long): CategoryDTO {
        val category = repository.findById(id)
            .orElseThrow { NoSuchElementException("Category not found with id: $id") }
        category.active = !category.active
        val updated = repository.save(category)
        return updated.toDTO()
    }
}
